RateLimiter/RateLimiterAsync.cs

```diff
@@ -0,0 +1,127 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RateLimiter
+{
+    public class RateLimiterAsync : IRateLimiterAsync
+    {
+        public static IRateLimiterAsync CreateInstance(IRateLimiter that)
+        {
+            return new RateLimiterAsync(that);
+        }
+
+        public static IRateLimiterAsync CreateInstance(
+            double ratePerSecond,
+            double maxAmount,
+            double initAmount = 0.0,
+            ITimeSource timeSource = default(ITimeSource)
+        )
+        {
+            var that = RateLimiterTokenBucket.CreateInstance(ratePerSecond, maxAmount, initAmount, timeSource);
+            return new RateLimiterAsync(that);
+        }
+
+        protected RateLimiterAsync(IRateLimiter rateLimiter)
+        {
+            _rateLimiter = rateLimiter;
+            _lastTask = null;
+        }
+
+        public bool ConsumeAmount(double numPermits = 1)
+        {
+            return _rateLimiter.ConsumeAmount(numPermits);
+        }
+
+        public long WaitTimeForAmount(double numPermits = 1)
+        {
+            return _rateLimiter.WaitTimeForAmount(numPermits);
+        }
+
+        /**
+         * If different processes are asking for different amounts, processes asking for larger amounts
+         * risk starvation.
+         *
+         * Our strategy to avoid starvation is a first-come -- first-serve policy: If we observe congestion,
+         * the first process to ask for an amount gets a reserve for that amount and waits until it's available.
+         *
+         * If a queue already exists, we're adding ourselves to the queue -- if there is no queue, we create
+         * one if we have to wait for our amount.
+         */
+        public async Task<bool> ConsumeAmountWhenAvailable(double amount = 1,
+            CancellationToken cancel = default(CancellationToken))
+        {
+            if (cancel.IsCancellationRequested)
+            {
+                return false;
+            }
+            
+            var ticket = default(Task<bool>);
+            lock (this)
```

Comment:
avoid using the following as lock objects:

this, as it might be used by the callers as a lock.

Reply:
good point, changed now